src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -100,49 +88,76 @@ private void StopRaisingEvents()
+            private bool _includeChildren;
 
-        // The background thread to use to monitor events
-        private Thread _watcherThread = null;
+            // The bitmask of events that we want to send to the user
+            private Interop.EventStream.FSEventStreamEventFlags _filterFlags;
 
-        // A reference to the RunLoop that we can use to start or stop a Watcher
-        private CFRunLoopRef _watcherRunLoop = IntPtr.Zero;
+            // The EventStream to listen for events on
+            private SafeEventStreamHandle _eventStream;
 
-        // Since all variables are IntPtr, Strings, or change references, using this in our locks
-        private readonly object _syncLock = new object();
+            // The background thread to use to monitor events
+            private Thread _watcherThread;
 
-        // Callback delegate for the EventStream events 
-        private Interop.EventStream.FSEventStreamCallback _callback = null;
+            // A reference to the RunLoop that we can use to start or stop a Watcher
+            private CFRunLoopRef _watcherRunLoop;
 
-        // Use an event to try to prevent StartRaisingEvents from returning before the
-        // RunLoop actually begins. This will mitigate a race condition where the watcher
-        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
-        private readonly AutoResetEvent _runLoopStartedEvent = new AutoResetEvent(false);
+            // Callback delegate for the EventStream events 
+            private Interop.EventStream.FSEventStreamCallback _callback;
 
-        private void CreateStreamAndStartWatcher()
-        {
-            Debug.Assert(_eventStream.IsInvalid);
-            Debug.Assert(_watcherRunLoop == IntPtr.Zero);
-            Debug.Assert(_callback == null);
+            // Use an event to try to prevent StartRaisingEvents from returning before the
+            // RunLoop actually begins. This will mitigate a race condition where the watcher
+            // thread hasn't completed initialization and stop is called before the RunLoop even starts.
+            private readonly AutoResetEvent _runLoopStartedEvent;
 
-            // Make sure we only do this if there is a valid directory
-            if (String.IsNullOrEmpty(_directory) == false)
+            // Token to monitor for cancellation requests, upon which processing is stopped and all
+            // state is cleaned up.
+            private readonly CancellationToken _cancellationToken;
+
+            internal RunningInstance(
+                FileSystemWatcher watcher,
+                string directory,
+                bool includeChildren,
+                Interop.EventStream.FSEventStreamEventFlags filter,
+                CancellationToken cancelToken)
             {
-                _fullDirectory = System.IO.Path.GetFullPath(_directory);
+                Debug.Assert(string.IsNullOrEmpty(directory) == false);
```

Comment:
Consider adding an assert here that !cancelToken.IsCancellationRequested.  This ctor is depending on a very specific pattern of cancellation, depending on the fact that the rest of the implementation controls cancellation and that cancellation can't be requested until after _watcherRunLoop has been initialized as part of the background operation.  If it were to be requested prior than that, things would break, as _watcherRunLoop would still be the default value, we'd attempt to stop something based on that default value, and then the real implementation would continue running unencumbered.